### PR TITLE
ci: Add nightly upload of wheels to Anaconda Cloud

### LINKED ISSUE #3011: Add nightly wheel upload to Anaconda Cloud
To move the idea of https://github.com/scientific-python/upload-nightly-action/issues/29 along sooner, I've made a [`scikit-hep-nightly-wheels` Anaconda Cloud organization](https://anaconda.org/scikit-hep-nightly-wheels) (only 3 GB of storage at the moment) and made an Awkward group there. 

image

Repurposing the workflow parts of https://github.com/matplotlib/matplotlib/blob/c2aa4ee1177452dc778813e3034c3f57d7db83d8/.github/workflows/nightlies.yml, originally added in https://github.com/matplotlib/matplotlib/pull/22733, on my fork I've created a similar workflow that is able to download the results of the [Packaging Tests workflow](https://github.com/scikit-hep/awkward/blob/dd4753b97949e4098aa6008a189624ba7a72fda6/.github/workflows/packaging-test.yml) and then upload to the `scikit-hep-nightly-wheels` Anaconda Cloud organization using

```yaml
      - name: Upload wheels to Anaconda Cloud as nightlies
        uses: scientific-python/upload-nightly-action@6e9304f7a3a5501c6f98351537493ec898728299 # 0.3.0
        with:
          artifacts_path: dist
          anaconda_nightly_upload_organization: scikit-hep-nightly-wheels
          anaconda_nightly_upload_token: ${{ secrets.ANACONDA_ORG_UPLOAD_TOKEN }}
```

successfully

image

```console
$ docker run --rm -ti python:3.11 /bin/bash
root@0487e039366a:/# python -m venv venv && . venv/bin/activate
(venv) root@0487e039366a:/# python -m pip --quiet install --upgrade pip wheel
(venv) root@0487e039366a:/# python -m pip install --upgrade --pre  --index-url https://pypi.anaconda.org/scikit-hep-nightly-wheels/simple --extra-index-url https://pypi.org/simple awkward
Looking in indexes: https://pypi.anaconda.org/scikit-hep-nightly-wheels/simple, https://pypi.org/simple
Collecting awkward
  Downloading awkward-2.6.1-py3-none-any.whl.metadata (7.0 kB)
Collecting awkward-cpp==29 (from awkward)
  Downloading awkward_cpp-29-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.2 kB)
Collecting fsspec>=2022.11.0 (from awkward)
  Downloading fsspec-2024.2.0-py3-none-any.whl.metadata (6.8 kB)
Collecting importlib-metadata>=4.13.0 (from awkward)
  Downloading importlib_metadata-7.0.1-py3-none-any.whl.metadata (4.9 kB)
Collecting numpy>=1.18.0 (from awkward)
  Downloading numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 1.1 MB/s eta 0:00:00
Collecting packaging (from awkward)
  Downloading packaging-23.2-py3-none-any.whl.metadata (3.2 kB)
Collecting zipp>=0.5 (from importlib-metadata>=4.13.0->awkward)
  Downloading zipp-3.17.0-py3-none-any.whl.metadata (3.7 kB)
Downloading awkward-2.6.1-py3-none-any.whl (749 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 750.0/750.0 kB 7.8 MB/s eta 0:00:00
Downloading awkward_cpp-29-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (706 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 706.6/706.6 kB 9.4 MB/s eta 0:00:00
Downloading fsspec-2024.2.0-py3-none-any.whl (170 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.9/170.9 kB 15.4 MB/s eta 0:00:00
Downloading importlib_metadata-7.0.1-py3-none-any.whl (23 kB)
Downloading numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 20.4 MB/s eta 0:00:00
Downloading packaging-23.2-py3-none-any.whl (53 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.0/53.0 kB 7.1 MB/s eta 0:00:00
Downloading zipp-3.17.0-py3-none-any.whl (7.4 kB)
Installing collected packages: zipp, packaging, numpy, fsspec, importlib-metadata, awkward-cpp, awkward
Successfully installed awkward-2.6.1 awkward-cpp-29 fsspec-2024.2.0 importlib-metadata-7.0.1 numpy-1.26.4 packaging-23.2 zipp-3.17.0
(venv) root@0487e039366a:/# python -m pip list
Package            Version
------------------ --------
awkward            2.6.1
awkward-cpp        29
fsspec             2024.2.0
importlib-metadata 7.0.1
numpy              1.26.4
packaging          23.2
pip                24.0
setuptools         65.5.0
wheel              0.42.0
zipp               3.17.0
(venv) root@0487e039366a:/#
```

If this sounds good, I'll PR what I have and I'll need at least 1 (perferably 2) maintainers to create Anaconda Cloud (https://anaconda.org/) accounts and make new API tokens with permissions described in https://scientific-python.org/specs/spec-0004/#process-for-adding-new-projects

image

### PR DESCRIPTION
**Pull request by matthewfeickert**
Resolves #3011 

* Add nightly schedule job for Packaging Tests workflow at 01:23 UTC to generate wheels.
* Add nightly wheels workflow that uses the `gh` command line tool to search and find the artifacts that were uploaded, downlod them, and then upload them to the [`scikit-hep-nightly-wheels` Anaconda Cloud org](https://anaconda.org/scikit-hep-nightly-wheels) using the [`scientific-python/upload-nightly-action` GitHub Action](https://github.com/scientific-python/upload-nightly-action).

I'm putting this in draft mode so that I can clean it up before this gets approved and merged.

### CODE REVIEW

**Review by jpivarski — APPROVED** [review]
I'm definitely in favor. This is a great thing to do; thanks!

Don't worry about the pypy test. It's allowed to fail as a non-required test to remind us about #2971.

**Comment by matthewfeickert** [comment]
> I'm definitely in favor. This is a great thing to do; thanks!

@jpivarski Great! I think beyond a review here, what would be good to discuss is also updates to the other workflows that are doing the wheel building to make sure that all desired wheels are built each night to get uploaded. At the moment https://github.com/scikit-hep/awkward/issues/3011#issuecomment-1933036482 makes it seems like only a subset are being built in the test workflows.

Also the

> If this sounds good, I'll PR what I have and I'll need at least 1 (perferably 2) maintainers to create Anaconda Cloud (https://anaconda.org/) accounts and make new API tokens with permissions described in https://scientific-python.org/specs/spec-0004/#process-for-adding-new-projects
>
> image

from https://github.com/scikit-hep/awkward/issues/3011#issue-2123979183 is needed before this can go in.

**Comment by jpivarski** [comment]
Is this only for compiled wheels or for pure Python packages as well? (I'm thinking about Uproot.) Pure Python packages are less in need of pre-built wheels because pip+git would work.

**Comment by matthewfeickert** [comment]
> Is this only for compiled wheels or for pure Python packages as well? (I'm thinking about Uproot.) Pure Python packages are less in need of pre-built wheels because pip+git would work.

Yeah, good point. It is totally up to us, as this is a package index that Scikit-HEP controls, so for single point of contact simplicity we could add pure-Python wheels, but for `pyhf`'s [HEAD of dependencies testing workflow](https://github.com/scikit-hep/pyhf/blob/fc5e83018d2a2c98156e5073a6e630793a1edd55/.github/workflows/dependencies-head.yml#L100-L120) simply doing

```console
...
python -m pip uninstall --yes uproot
python -m pip install --upgrade git+https://github.com/scikit-hep/uproot5.git
```

has worked fine. So I'd defer to you and the rest of the `uproot` team on this.

**Comment by matthewfeickert** [comment]
@jpivarski Okay, I cleaned up a few things in the PR and gave it nicer commit messages for the history. Once you have done the following:

* [x] Created an account on [Anaconda Cloud](https://anaconda.org/) (oh, is this you: https://anaconda.org/jpivarski ?)
* [x] Told me the name of your account
* [x] Made a new Anaconda Cloud API token with permissions described in https://scientific-python.org/specs/spec-0004/#process-for-adding-new-projects
* [x] Saved the API token as a repo GitHub Actions secret named `ANACONDA_ORG_UPLOAD_TOKEN`
* [x] Verified that I have added you to the Awkward team account on https://anaconda.org/scikit-hep-nightly-wheels

This can get merged in and should start working.

https://github.com/scikit-hep/awkward/blob/dd4753b97949e4098aa6008a189624ba7a72fda6/.github/workflows/packaging-test.yml only builds a handful of wheels, and so isn't quite providing true nightly wheels for all targets yet, but this is a solid first step and I think the Awkward team can exapnd on this to meet whatever goals they want from here.

**Comment by agoose77** [comment]
Right now we don't publish any wheels built on main, unless I am misremembering something! We do build PR wheels, which you rightly spot in `packaging-test.yml`. 

Given that we test quite a lot of platform-specific stuff in our main PR test suite, I'd be tempted to change `packaging-test.yml` to test *only* `main` via a CRON job (or maybe on PUSH). @jpivarski thoughts?

File: .github/workflows/packaging-test.yml
Comment (matthewfeickert) [inline .github/workflows/packaging-test.yml:7]: @agoose77 RE: https://github.com/scikit-hep/awkward/pull/3012#issuecomment-1934159262, yeah as there is no testing on `main` yet, this is why I added the daily CRON bit here. Are you suggesting to no longer test on PRs, but just this bit?

Comment (matthewfeickert) [reply .github/workflows/packaging-test.yml:7]: So something like

```yaml
on:
  push:
    branches:
    - main
  # Run daily at 1:23 UTC
  schedule:
    - cron: '23 1 * * *'
  workflow_dispatch:
```

?

Comment (agoose77) [reply .github/workflows/packaging-test.yml:7]: I am on mobile and coming down with a cold, so not super speedy of thought right now! What I'm proposing is to only run full blown packaging tests on main, if we want to save the cycles associated with it.

If Jim doesn't feel the need to do this, then we should definitely add the cron as you have!

Comment (matthewfeickert) [reply .github/workflows/packaging-test.yml:7]: Ah gotcha. Yeah, this is what I did for `matplotlib` (https://github.com/matplotlib/matplotlib/blob/c2aa4ee1177452dc778813e3034c3f57d7db83d8/.github/workflows/nightlies.yml) so I'm happy to change it to be `main` and `push` based.

Comment (jpivarski) [reply .github/workflows/packaging-test.yml:7]: We absolutely need to keep testing commits on PR branches, and it is the case that `main` gets tested after a merge. (I don't know which part of the CI does this, but I've observed that it happens every time, with the way we have things set up.)

I guess it's advantageous to test on a regular interval, to catch when third-party updates break things, and daily is the right cadence for that. So sure, let's add daily testing on `main`.

This is in addition to (or part of) the process that makes daily artifacts for Anaconda Cloud?

File: .github/workflows/packaging-test.yml
Comment (matthewfeickert) [inline .github/workflows/packaging-test.yml]: Now running on pushes to `main` only (beyond workflow dispatch).

File: .github/workflows/upload_nightly_wheels.yml
Comment (matthewfeickert) [inline .github/workflows/upload_nightly_wheels.yml]: Now selecting the `push` events that were on `main`, which comes from runs set off by merged PRs.

**Review by matthewfeickert — COMMENTED** [review]
@agoose77 here's updated logic. @jpivarski These revisions also look good to you?

File: .github/workflows/packaging-test.yml
Comment (matthewfeickert) [reply .github/workflows/packaging-test.yml:7]: Oh sorry, @jpivarski, async passing and missed this. Let me read and revert as needed.

Comment (agoose77) [reply .github/workflows/packaging-test.yml:7]: @jpivarski I don't think we explicitly test main, apart from docs right now. Our PR tests are evaluated against the merge commit with main due to the checkout action (I am pretty sure), so it's fairly safe.

Comment (matthewfeickert) [reply .github/workflows/packaging-test.yml:7]: >  it is the case that main gets tested after a merge. (I don't know which part of the CI does this, but I've observed that it happens every time, with the way we have things set up.)
> 
> I guess it's advantageous to test on a regular interval, to catch when third-party updates break things, and daily is the right cadence for that. So sure, let's add daily testing on main.

If there is already instances in which there are runs that produce wheels on `main` then that's really enough, and the nightly wheels are more about capturing the state of dev at `HEAD`, so let me revert what I just added and then look at what workflows run on pushes to `main` from merges.

Comment (agoose77) [reply .github/workflows/packaging-test.yml:7]: Actually no wheels are built an published on main. We do build the single platform for docs, but that's not using cibuildwheel.

Comment (matthewfeickert) [reply .github/workflows/packaging-test.yml:7]: Yeah, looking at `push` events on `main` (https://github.com/scikit-hep/awkward/actions?query=event%3Apush+branch%3Amain) this shows what @agoose77 just mentioned

image

Comment (jpivarski) [reply .github/workflows/packaging-test.yml:7]: I was referring to testing, not wheel-publishing. I've seen tests on `main` after each PR merge, but I take it that the wheel-building/publishing would be a new step.

File: .github/workflows/packaging-test.yml
Comment (matthewfeickert) [inline .github/workflows/packaging-test.yml]: Given https://github.com/scikit-hep/awkward/pull/3012#discussion_r1483499237 kept PRs and schedule, but added push events. Can be removed.

File: .github/workflows/packaging-test.yml
Comment (jpivarski) [reply .github/workflows/packaging-test.yml:7]: Ah, you were only ever talking about wheel building/publishing. Yes, we can add that on `main` as a new feature.

**Review by matthewfeickert — COMMENTED** [review]
@jpivarski @agoose77 Reverted things, but added runs on `push` events on `main` until you can tell me if you want this or just the `schedule` events added here. At the moment, the logic for downloading is only checking for `schedule` events, but it could look for `schedule` or `push` and just get whatever is the latest.

File: .github/workflows/packaging-test.yml
Comment (agoose77) [reply .github/workflows/packaging-test.yml:7]: Just to make the point for clarity (I am sure we are all on the same page), we don't really need to test main given that we're testing the merge commit. I'm pretty certain that's bullet proof.

As such, we don't need the packaging tests in PRs, as we build wheels for all (?) platforms. However, the packaging test ensures our cibuildwheel works, which isn't otherwise tested until deployment. 

But we don't want to build nighties on every pr, just the ones that made it to main. So to me, that means building main.

**Review by jpivarski — APPROVED** [review]
I'm still in favor of this; it looks like a good thing to do. I wonder what motivates the odd time, though. It's late enough to be evening, even in the U.S. (We'll soon have a new collaborator in Mountain time, but it's still okay for that—not likely to take away task-runners when we want feedback on a PR.)

**Comment by matthewfeickert** [comment]
> I wonder what motivates the odd time, though

Ah that is easy to explain. Running jobs at 0:00 UTC gets slowed way down because _many_ people are using that as a default and so GitHub Actions gets swamped with runner usage right then. Running at 1:23 UTC then gives you a little offset (and the numbers are just silly placeholders) and then running the uploader at 2:34 UTC means that you give it a huge amount of time to finish in the event that anything ever goes wrong and things get delayed by many minutes.

File: .github/workflows/packaging-test.yml
Comment (matthewfeickert) [reply .github/workflows/packaging-test.yml:7]: > But we don't want to build nighties on every pr, just the ones that made it to main. So to me, that means building main.

Yes, this is what we did for `matplotlib`. But at the moment the _full_ wheel building, as opposed to the subset of wheel build tests, is only run on deployment.

https://github.com/scikit-hep/awkward/blob/dd4753b97949e4098aa6008a189624ba7a72fda6/.github/workflows/deploy.yml#L1-L11

So what would be better would be to split the wheel building out into a seperate workflow and then have that workflow get called during publishing. Idk if you want to do that in a seperate PR though.

> Just to make the point for clarity (I am sure we are all on the same page), we don't really need to test main given that we're testing the merge commit. I'm pretty certain that's bullet proof.

Right. So I'm going to revert the last commit to where things were before we got talkinga bout this.

File: .github/workflows/packaging-test.yml
Comment (matthewfeickert) [reply .github/workflows/packaging-test.yml]: Removing.

**Comment by matthewfeickert** [comment]
To make testing of this easier in advance of merge, I just triggered a workflow dispatch run of `.github/workflows/packaging-test.yml` on Awkward which finished and then I pushed a change that looks for both `schedule` and `workflow_dispatch` event triggers. So this shell script

```bash
#!/bin/bash

rm -rf awkward-cpp-wheels-* awkward-wheel dist runs.json

PROJECT_REPO="scikit-hep/awkward"
BRANCH="main"
WORKFLOW_NAME="packaging-test.yml"
ARTIFACT_PATTERN="awkward*wheel*"  # awkward-wheel and awkward-cpp-wheels-*

gh run --repo "${PROJECT_REPO}" \
   list --branch "${BRANCH}" \
        --workflow "${WORKFLOW_NAME}" \
        --json event,status,conclusion,databaseId > runs.json
RUN_ID=$(
  jq --compact-output \
    '[
      .[] |
      # Filter on "schedule" and "workflow_dispatch" events to main (nightly build) ...
      select(.event == "schedule" or .event == "workflow_dispatch") |
      # that have completed successfully ...
      select(.status == "completed" and .conclusion == "success")
     ] |
    # and get ID of latest build of wheels.
    sort_by(.databaseId) | reverse | .[0].databaseId' runs.json
)
gh run --repo "${PROJECT_REPO}" view "${RUN_ID}"
gh run --repo "${PROJECT_REPO}" \
   download "${RUN_ID}" --pattern "${ARTIFACT_PATTERN}"

mkdir dist
mv ${ARTIFACT_PATTERN}/*.whl dist/
ls -l dist/
```

is the logic I implemented in the PR. If you run it

```console
$ bash debug.sh 

✓ main Packaging Tests · 7836312788
Triggered via workflow_dispatch about 16 minutes ago

JOBS
✓ Build awkward-cpp: windows-latest in 6m41s (ID 21383661203)
✓ Build awkward in 45s (ID 21383661437)
✓ Build awkward-cpp: macos-latest in 8m33s (ID 21383661690)
✓ Build awkward-cpp: ubuntu-latest in 7m1s (ID 21383661957)

ARTIFACTS
awkward-sdist
awkward-wheel
awkward-cpp-sdist
awkward-cpp-wheels-windows-latest
awkward-cpp-wheels-ubuntu-latest
awkward-cpp-wheels-macos-latest

For more information about a job, try: gh run view --job=<job-id>
View this run on GitHub: https://github.com/scikit-hep/awkward/actions/runs/7836312788
total 3844
-rw-r--r-- 1 feickert feickert  761870 Feb  8 15:57 awkward-2.6.1-py3-none-any.whl
-rw-r--r-- 1 feickert feickert  706425 Feb  8 15:57 awkward_cpp-29-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
-rw-r--r-- 1 feickert feickert  707671 Feb  8 15:57 awkward_cpp-29-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
-rw-r--r-- 1 feickert feickert 1214783 Feb  8 15:57 awkward_cpp-29-cp38-cp38-macosx_10_9_universal2.whl
-rw-r--r-- 1 feickert feickert  533474 Feb  8 15:57 awkward_cpp-29-cp39-cp39-win_amd64.whl
```

it gets the wheels that were just built. :+1:  That's the whole wheel gathering logic that I added and then the rest of the PR is just the uploader action.

**Comment by matthewfeickert** [comment]
So as the wheels all have the same name each time (they don't have vcs information associated with them) they _currently_ need some manual rerunning to deal with https://github.com/Anaconda-Platform/anaconda-client/issues/702, but after that, the workflow runs fine.

[image](https://anaconda.org/scikit-hep-nightly-wheels/awkward/files)


[image](https://anaconda.org/scikit-hep-nightly-wheels/awkward-cpp/files)

**Pull request merged**
